src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs

```diff
@@ -72,32 +75,76 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(string databaseProv
-
         try
         {
             var selectCommand = connection.CreateCommand();
             selectCommand.CommandText = selectBuilder.ToSqlString();
 
             using var result = await selectCommand.ExecuteReaderAsync();
-
-            // at this point the query succeeded and the table exists
-            return DbConnectionValidatorResult.DocumentFound;
+            if (!isDefaultShell)
+            {
+                // The 'Document' table exists.
+                return DbConnectionValidatorResult.DocumentTableFound;
+            }
+
+            var requiredColumnsCount = Enumerable.Range(0, result.FieldCount)
+                .Select(result.GetName)
+                .Where(c => _requiredDocumentTableColumns.Contains(c))
+                .Count();
+
+            if (requiredColumnsCount != _requiredDocumentTableColumns.Length)
+            {
+                // The 'Document' table exists with another schema.
+                return DbConnectionValidatorResult.DocumentTableFound;
+            }
         }
         catch
         {
-            // at this point we know that the document table does not exist
+            // The 'Document' table does not exist.
+            return DbConnectionValidatorResult.DocumentTableNotFound;
+        }
+
+        selectBuilder = GetSelectBuilderForShellDescriptorDocument(tablePrefix, providerName);
+        try
+        {
+            var selectCommand = connection.CreateCommand();
+            selectCommand.CommandText = selectBuilder.ToSqlString();
 
-            return DbConnectionValidatorResult.DocumentNotFound;
+            using var result = await selectCommand.ExecuteReaderAsync();
+            if (!result.HasRows)
+            {
+                // The 'Document' table exists with no 'ShellDescriptor' document.
+                return DbConnectionValidatorResult.ShellDescriptorDocumentNotFound;
+            }
+        }
+        catch
+        {
         }
+
+        // The 'Document' table exists.
+        return DbConnectionValidatorResult.DocumentTableFound;
     }
 
     private ISqlBuilder GetSelectBuilderForDocumentTable(string tablePrefix, DatabaseProviderName providerName)
     {
         var selectBuilder = GetSqlBuilder(providerName, tablePrefix);
 
         selectBuilder.Select();
-        selectBuilder.AddSelector("*");
+        selectBuilder.Selector("*");
+        selectBuilder.Table(_tableNameConvention.GetDocumentTable());
+        selectBuilder.Take("1");
+
+        return selectBuilder;
+    }
+
+    private ISqlBuilder GetSelectBuilderForShellDescriptorDocument(string tablePrefix, DatabaseProviderName providerName)
+    {
+        var selectBuilder = GetSqlBuilder(providerName, tablePrefix);
+
+        selectBuilder.Select();
+        selectBuilder.Selector("*");
         selectBuilder.Table(_tableNameConvention.GetDocumentTable());
+        selectBuilder.WhereAnd($"Type = '{typeof(ShellDescriptor).FullName}, {typeof(ShellDescriptor).Assembly.GetName().Name}'");
```

Comment:
Isn't there a helper in yessql to build this? The logic might change.

Reply:
Didn't find it but will check it again

Reply:
Okay I think I found it => `TypeService`